### PR TITLE
use template to support reserved contract args

### LINKED ISSUE #136: add contract names args
**Brief of your idea**
reserved contract needs contract_names field.

**What could be better if XuperChain using your idea**
Describe the benefits of your idea

**Are there any side effects?**
Describe what might be worse.

**Additional information**
If possible, include a minimal, complete, and verifiable example to help us identify the issue.

### PR DESCRIPTION
**Pull request by qizheng09**
fix #136

### CODE REVIEW

File: ledger/types.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package ledger
+
+import (
+	"github.com/xuperchain/xuperunion/pb"
+)
+
+type InvokeRequest struct {
```
Comment (yucaowang) [inline ledger/types.go]: Always leave comments on public structs/functions.

File: ledger/types.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	ModuleName   string            `json:"module_name"`
+	ContractName string            `json:"contract_name"`
+	MethodName   string            `json:"method_name"`
+	Args         map[string]string `json:"args"`
+}
+
+func invokeRequestFromJson2Pb(jsonRequest []InvokeRequest) ([]*pb.InvokeRequest, error) {
```
Comment (yucaowang) [inline ledger/types.go]: Json should be JSON or you will fail in golint check

File: utxo/reserved_contract.go [hunk 3]
```diff
@@ -2,12 +2,24 @@ package utxo
...
 
+// reservedArgs used to get contractnames from InvokeRPCRequest
+type reservedArgs struct {
+	ContractNames string
+}
+
+func genArgs(req []*pb.InvokeRequest) *reservedArgs {
+	ra := &reservedArgs{}
+	for _, v := range req {
+		ra.ContractNames += v.GetContractName() + ","
```
Comment (yucaowang) [inline utxo/reserved_contract.go]: this will add a comma at the end of the serialized string?

File: utxo/utxo.go [hunk 4]
```diff
@@ -740,7 +739,9 @@ func (uv *UtxoVM) PreExec(req *pb.InvokeRPCRequest, hd *global.XContext) (*pb.In
 			// FIXME zq @icexin need to return contract not found error
 			uv.xlog.Error("PreExec NewContext error", "error", err,
 				"contractName", tmpReq.GetContractName())
-			if i < len(reservedRequests) && err.Error() == "Key not found" {
+			if i < len(reservedRequests) && strings.HasSuffix(err.Error(), "not found") {
+				request := *tmpReq
```
Comment (yucaowang) [inline utxo/utxo.go]: why not using tmpReq direct to append into request?

Comment (qizheng09) [reply utxo/utxo.go]: because  the last `tmpReq` will overwrite the ahead

**Pull request merged**